modules/control-utility/src/main/java/org/apache/ignite/internal/commandline/CommandInvoker.java

```diff
@@ -0,0 +1,316 @@
+
+    /** Client configuration. */
+    private GridClientConfiguration clientCfg;
+
+    /** @param cmd Command to execute. */
+    public CommandInvoker(Command<A, ?> cmd, A arg, GridClientConfiguration clientCfg) {
+        this.cmd = cmd;
+        this.arg = arg;
+        this.clientCfg = clientCfg;
+    }
+
+    /**
+     * Actual command execution with verbose mode if needed.
+     * Implement it if your command supports verbose mode.
+     *
+     * @param log Logger to use.
+     * @param verbose Use verbose mode or not
+     * @return Result of operation (mostly usable for tests).
+     * @throws Exception If error occur.
+     */
+    public <R> R invoke(IgniteLogger log, boolean verbose) throws Exception {
+        try (GridClient client = startClient(clientCfg)) {
+            String deprecationMsg = cmd.deprecationMessage(arg);
+
+            if (deprecationMsg != null)
+                log.warning(deprecationMsg);
+
+            R res;
+
+            if (cmd instanceof LocalCommand)
+                res = ((LocalCommand<A, R>)cmd).execute(client, arg, log::info);
+            else if (cmd instanceof ComputeCommand) {
+                GridClientCompute compute = client.compute();
+
+                Map<UUID, GridClientNode> nodes = compute.nodes().stream()
+                    .collect(toMap(GridClientNode::nodeId, n -> n));
+
+                ComputeCommand<A, R> cmd = (ComputeCommand<A, R>)this.cmd;
+
+                Collection<UUID> cmdNodes = cmd.nodes(nodes, arg);
+
+                if (cmdNodes == null)
+                    cmdNodes = singleton(defaultNode(client, clientCfg).nodeId());
+
+                for (UUID id : cmdNodes) {
+                    if (!nodes.containsKey(id))
+                        throw new IllegalArgumentException("Node with id=" + id + " not found.");
+                }
+
+                Collection<GridClientNode> connectable = F.viewReadOnly(
+                    cmdNodes,
+                    nodes::get,
+                    id -> nodes.get(id).connectable()
+                );
+
+                if (!F.isEmpty(connectable))
+                    compute = compute.projection(connectable);
+
+                res = compute.execute(cmd.taskClass().getName(), new VisorTaskArgument<>(cmdNodes, arg, false));
+
+                cmd.printResult(arg, res, log::info);
+            }
+            else
+                throw new IllegalArgumentException("Unknown command type: " + cmd);
+
+            return res;
+        }
+        catch (Throwable e) {
+            log.error("Failed to perform operation.");
+            log.error(CommandLogger.errorMessage(e));
+
+            throw e;
+        }
+    }
+
+    /** */
+    public boolean prepare(IgniteLogger log) throws Exception {
+        if (!(cmd instanceof PreparableCommand))
+            return true;
+
+        try (GridClient client = startClient(clientCfg)) {
+            return ((PreparableCommand<A, ?>)cmd).prepare(client, arg, log::info);
+        }
```

Comment:
Are you sure that it is not required here to wrap exception into `GridClientException`?

Reply:
Yes. Method changed to `public boolean prepare(GridClient cli, A arg, Consumer<String> printer) throws GridClientException;`.